clang: -Wuninitialized false positive when accessing a virtual base class member in a member initializer of an abstract class constructor

The code

```c++
struct C1 {
  C1() {}

  int foo() { return 0; }
};

struct C2 : virtual C1 {
  int i;

  C2() : i(this->foo()) {}

  virtual void makeAbstract() = 0;
};

struct C3 : C2 {
  void makeAbstract() {}
};

int main() { C3 c; }
```

will cause the warning

```
file.cpp:10:18: warning: base class 'C1' is uninitialized when used here to access 'C1::foo' [-Wuninitialized]
   10 |   C2() : i(this->foo()) {}
      |                  ^
```

even though C1, being a base class of C2, should always be initialized before the non-static data members of C2 are initialized (according to my understanding of [class.base.init] in the C++11 standard).

Note that making C2 non-abstract will remove the warning, even though it should not change anything.

Also, changing the C2 constructor to `C2() : C1(), i(this->foo()) {}` will remove the warning, even though the call to `C1()` will always be ignored because C2 is abstract.

--- a/clang/lib/Sema/SemaDeclCXX.cpp
+++ b/clang/lib/Sema/SemaDeclCXX.cpp
@@ -4158,8 +4158,13 @@ namespace {
     }
 
     llvm::SmallPtrSet<QualType, 4> UninitializedBaseClasses;
-    for (const auto &I : RD->bases())
+    for (const auto &I : RD->bases()) {
+      // Virtual bases are initialized from the most derived class, so an
+      // abstract base class constructor can assume it to be initialized.
+      if (I.isVirtual() && RD->isAbstract())
+        continue;
       UninitializedBaseClasses.insert(I.getType().getCanonicalType());
+    }
 
     if (UninitializedFields.empty() && UninitializedBaseClasses.empty())
       return;

--- a/clang/test/SemaCXX/uninitialized.cpp
+++ b/clang/test/SemaCXX/uninitialized.cpp
@@ -1411,6 +1411,8 @@ struct B : public A {
 
   B(int (*)[4]) : A(foo()) {}
   // expected-warning@-1 {{base_class_access::A' is uninitialized when used here to access 'base_class_access::A::foo'}}
+
+  virtual void f() = 0;
 };
 
 struct C {
@@ -1426,6 +1428,22 @@ struct D : public C, public A {
 
   D(int (*)[4]) : C(foo()) {}
   // expected-warning@-1 {{base_class_access::A' is uninitialized when used here to access 'base_class_access::A::foo'}}
+};
+
+struct E : public virtual A {
+  E(int (*)[3]) : A(i) {}
+  // expected-warning@-1 {{base class 'base_class_access::A' is uninitialized when used here to access 'base_class_access::A::i'}}
+
+  E(int (*)[4]) : A(foo()) {}
+  // expected-warning@-1 {{base_class_access::A' is uninitialized when used here to access 'base_class_access::A::foo'}}
+};
+
+// Accessing virtual base class members from an abstract class is fine.
+struct F : public virtual A, public C {
+  F(int (*)[3]) : C(i) {}
+  F(int (*)[4]) : C(foo()) {}
+
+  virtual void f() = 0;
 };
 
 }

